mtags-interfaces/src/main/java/scala/meta/pc/SymbolSearch.java

```diff
@@ -19,6 +19,11 @@
      */
     List<Location> definition(String symbol);
 
+    /**
+     * Returns the file where the symbol is defined, if any.
+     */
+    Optional<VirtualFile> definitionSource(String symbol);
```

Comment:
I'm wondering if we should change this API to directly return the toplevels 🤔 
```suggestion
    List<String> definitionSourceToplevels(String symbol);
```
This would allow the Metals server to potentially cache the results between different invocations. The `Mtags.toplevels` method is on the other hand very fast (~900k lines/sec) I have a feeling that a lot of `match (exhaustive)` completions are on the same types (for example `scala/Option#`).

Reply:
> I'm wondering if we should change this API to directly return the toplevels 🤔

Sure! That makes more sense, and now we can remove the java version of `VirtualFile` and `VirtualFileImpl` 😄

Reply:
Thank you so much for your input @olafurpg !

I just added a straightforward implementation of the cache for `definitionSourceToplevels` in https://github.com/scalameta/metals/pull/1174/commits/a70886e97250f94890b922cb89662fb911f3eaf5

Though it definitely makes sense to introduce cache for `definitionSourceToplevels` :+1:, I have one concern about the cache invalidation:
In https://github.com/scalameta/metals/pull/1174/commits/a70886e97250f94890b922cb896 the cache in `MetalsSymbolSearch` will be cleared only when the [workspace is updated](https://github.com/scalameta/metals/blob/2475a952a4eaf7f3b5e41246e091c3757f244908/metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala#L390-L407).
If we add some new subclasses to a `sealed abstract class X`, metals server would be unable to reflect the changes until the workspace is updated.

---

To avoid this, it might be better to do:

- cache `definitionSourceToplevels` result by the symbol of definition's (absolute) path
  - `private val cache = new TrieMap[AbsolutePath, ju.List[String]]()`
- Everytime the project is compiled, clear the cache for `definitionSourceToplevels` at [Compilers#didCompile](https://github.com/scalameta/metals/blob/2475a952a4eaf7f3b5e41246e091c3757f244908/metals/src/main/scala/scala/meta/internal/metals/Compilers.scala#L109-L122) before restart the PresentationCompiler instance.
  - so the metals server can reflect the change to the completion items.
  - to do this, we have to add cache cleanup method to `SymbolSearch`
- The cache items we have to clear are the items whose key (`AbsolutePath`) is located under the project's workspace.
  - We don't need to clear the symbol cache from the 3rd party definitions, because they won't change without updating the workspace.

What do you think about this?

Reply:
Cache invalidation is a good question. Can we maybe reuse the toplevels indexing for WorkspaceSymbolProvider? That index is already updated on file save.